changelog_unreleased/typescript/10187.md

```diff
@@ -0,0 +1,24 @@
+#### Fix formatting for ternary operators with `as`. (#10187 by @sosukesuzuki)
+
+<!-- prettier-ignore -->
+```ts
+// Input
+const ternaryAndTypescriptAs = (
+  callNode.parent?.type === AST_NODE_TYPES.ChainExpression
+    ? callNode.parent.parent
+    : callNode.parent
+) as TSESTree.BinaryExpression;
+
+// Prettier stable
+const ternaryAndTypescriptAs = (callNode.parent?.type ===
+AST_NODE_TYPES.ChainExpression
+  ? callNode.parent.parent
+  : callNode.parent) as TSESTree.BinaryExpression;
+
+// Prettier main
+const ternaryAndTypescriptAs = (
+  callNode.parent?.type === AST_NODE_TYPES.ChainExpression
+    ? callNode.parent.parent
+    : callNode.parent
+) as TSESTree.BinaryExpression;
```

Comment:
Do you think we should consistent these outputs? All different.

**Prettier pr-10187**
[Playground link](https://deploy-preview-10187--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuc0DOMAE8BOUCGOAngIJQAmAKkQA5xpg4CWNMJamAvJgBQA6UTJjD4ANqIByEcnAB0NQghgB+WTFpwunbiQDKlAPoSA8gBEAogcoBNAArndsgMIALfEyjmAHjRz00TNACQkLKwmKS0nIKfrDyirDBIUjh4lIy8bEwAgCUmPgclLoOlH5yAEIehETevv6BUADcAgKQUBjYcHjVZFQaDMys7Fy8SSJpUZlKqup0Wjr6RmaWNvaOru6ePn5oAUGCIWHjkRkxSlOJB0Ipx+nRCdlQObJFJWWylQTEtTt7TS1QNodXBfUgUah0AYsNgcbj8A63SZnWAzDTzTB6QwmCxWOwOZxuDw-er7EKYI4RO4XGDUpLXVIne5ZXIAbVe+nen2qxN2DQAuo0QAAaEAQVgNNDIUCEHAQADutgSkpQYjl+CIkpFACMcPgwABrOAwXQKMAeADmyBgOAArnARXAALZauDkGTkAAy+Cg5pt+HNcAAYhAcI78DAYBbkCB8DaYBBhSAXDBHaIAOouJjwNCmuC6BABSMANyzRGjYF2iY8aC6MFsuvNYeQADMxDWRQArNBecq6g1G3T4R1wD0eOAttv2kBdry6C2iOAARRtEHgE9E7ZAMRrOGjs3ojGhid8HhgaaY5BgLmQAA4AAwi3wQGtp3U0aN1HdF8cigCOK-gesxWVGM0AAWigOBXVdRM-H-Jg-Hrf0myQVsNynGtHSYK1bQw+clwA8dUMnEUYHwLVz0va8kAAJlI3UmFEC0nAgR0UJAegAFZExtGtKHI5U0M3Is7QASQoJRdEPIYKF0dQF3XGsAF8lKAA)
```sh
--parser typescript
```

**Input:**
```tsx
const ternaryAndTypescriptAs = (
  callNode.parent?.type === AST_NODE_TYPES.ChainExpression
    ? callNode.parent.parent
    : callNode.parent
) as TSESTree.BinaryExpression;

const ternaryAndTypescriptAs = (
  callNode.parent?.type === AST_NODE_TYPES.ChainExpression
    ? callNode.parent.parent
    : callNode.parent
).TSESTree.BinaryExpression;

const ternaryAndTypescriptAs = (
  callNode.parent?.type === AST_NODE_TYPES.ChainExpression
    ? callNode.parent.parent
    : callNode.parent
)[TSESTree.BinaryExpression];
```

**Output:**
```tsx
const ternaryAndTypescriptAs = (
  callNode.parent?.type === AST_NODE_TYPES.ChainExpression
    ? callNode.parent.parent
    : callNode.parent
) as TSESTree.BinaryExpression;

const ternaryAndTypescriptAs = (callNode.parent?.type ===
AST_NODE_TYPES.ChainExpression
  ? callNode.parent.parent
  : callNode.parent
).TSESTree.BinaryExpression;

const ternaryAndTypescriptAs = (callNode.parent?.type ===
AST_NODE_TYPES.ChainExpression
  ? callNode.parent.parent
  : callNode.parent)[TSESTree.BinaryExpression];

```

Reply:
Yes, I'll work on it after merge this.

Reply:
Even though I like the new output, but I think we never break after `(`, I think we should choose one style for these cases, because this case format exactly like the computed one before. If you think they are all broken, you should fix both. Otherwise they are all not broken.

Reply:
I agree with you. Should I fix it on this PR? I think to support `MemberExpression` and `CallExpression` make logics bit complex. So I want to keep simple this PR and create new PR.

(`CallExpression` example:
```js
const ternaryAndTypescriptAs = (callNode.parent?.type ===
AST_NODE_TYPES.ChainExpression
  ? callNode.parent.parent
  : callNode.parent)(TSESTree.BinaryExpression);
```
)

Reply:
I think should done in this PR, it's like introducing a  third format, let's see if current format suit other cases .

If you want keep this one simple, we can do others without this `as` first.

Reply:
>  I think we never break after (

It's done at least for `await`:

**Prettier 2.2.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBDAzgTymABAMwFccYBLaAgCgEpdgAdKXXSKdGXAI1QC9cBeXJVQB3VKQ74IEAGLS5s+UsUqFa6TWoA6bgCcAQqgNG9h47tNGz1ozQDcjAL4gANCAgAHMtHTJQR3QgRAAUjBF8UVAAbMUxfN05dVDAAazgYAGUPZNIoAHNkGF1CODc4AFtOOAATapqAGVR8wlQ8uDldctQYMnzkNEIYCFcQAAsYcqiAdVGJOHRssDgM8IlSADcJTH6wdHiQXPQ4XRhgpLyu5Hxoo7cAK3QAD30k1PSM1HK4ety4K5vSiAHo8Mrk8lE4ABFQgQeD-KK3EDZXRHXT9GCYDzzMC6UheEYeXGwKakaowUbIAAcAAY3ISIEcpkkPP1CfNjus-m4AI4w+BnTwRNDoAC0UDgNRqI10cF5pBlZ1alyQ1wRgKO5VIhWK6rBEOhsL+KoBbhgqE4JLJFKQACZTUlSFEwQBhCDlZUgeYAVhGhCOABVzRFVYj1iUAJJQOqwDI4vEwACCUYyGIh8KOjkcQA)

```tsx
async function f() {
  const baz = (
    await fooFooFooFooFooFooFooFooFooFooFoo()
  ).barBarBarbarBarBarbarBarBarBarBar();
}

```

Reply:
I think extra indention for ternary looks good, my point is we should apply to these metioned cases too, better in one PR, so we can compare different types of Node.

Reply:
This one #7564 kind of related.

Reply:
@fisker @thorn0 I've updated this PR for consistent for `TSAsExpression`, `MemberExpression`, `CallExpression`. Please review again.